tun/tun_linux.go

```diff
@@ -399,6 +491,50 @@ func (tun *NativeTun) Close() error {
 	return err2
 }
 
+func (tun *NativeTun) BatchSize() int {
+	return tun.batchSize
+}
+
+const (
+	// TODO: support TSO with ECN bits
+	tunOffloads = unix.TUN_F_CSUM | unix.TUN_F_TSO4 | unix.TUN_F_TSO6
+)
+
+func (tun *NativeTun) initFromFlags(name string) error {
+	sc, err := tun.tunFile.SyscallConn()
+	if err != nil {
+		return err
+	}
+	if e := sc.Control(func(fd uintptr) {
+		var (
+			ifr *unix.Ifreq
+		)
+		ifr, err = unix.NewIfreq(name)
+		if err != nil {
+			return
+		}
+		err = unix.IoctlIfreq(int(fd), unix.TUNGETIFF, ifr)
+		if err != nil {
+			return
+		}
+		got := ifr.Uint16()
+		if got&unix.IFF_VNET_HDR != 0 {
+			err = unix.IoctlSetInt(int(fd), unix.TUNSETOFFLOAD, tunOffloads)
+			if err != nil {
+				return
+			}
+			tun.vnetHdr = true
+			tun.batchSize = conn.DefaultBatchSize
+		} else {
+			tun.batchSize = 1
```

Comment:
Maybe we should warn here? Seems like that'd be a bug or indicative of an old kernel or something, right?

Reply:
This can be reached from `CreateTUN()` [1], `CreateTunFromFile()` [2], and `CreateUnmonitoredTUNFromFD()` [3].

1 is the common path used by wireguard-go. It sets `IFF_VNET_HDR` and returns an error if it fails.

2 is used by wireguard-go when the user supplies their own FD via env var (WG_TUN_FD). It assumes the user has already performed the `TUNSETIFF` `ioctl()`, which can only be done once. The tun status hack won't work if the user didn't set `IFF_VNET_HDR` prior.

3 is used by wireguard-android.  We know that recent stock Android blocks `IFF_VNET_HDR` via SELinux policy, so the tun status hack can't work there (for now).

In the current state we could add a log, but returning an error would break 3 and probably some usages of 2.

Reply:
Ah, darn, right. (2) isn't an issue -- that environment variable is an internal design detail to deal with daemonization on fork()-less Go. But (3) certainly is. Too bad about IFF_VNET_HDR on Android... I imagine the Tailscale crew must have some old Google colleagues to poke about that.